TickCounter class refactoring

alexandre-lecoq opened the pull request.

maehne commented:
Thanks for the PR @alexandre-lecoq! Could you please provide a bit more information, what you changed and why to facilitate its review?

alexandre-lecoq commented:
Hello,
While investigating why the tick rate was slow in simulations, i stumbled on this class and noticed that propagationCompleted() which is often called was doing lots of work, so I made propagationCompleted() lighter, and also removed the arrays, and tried to make the code easier to understand at first glance.

maehne reviewed (approved):
Thanks @alexandre-lecoq for the PR and explanation! Looks good to me, but I suggest that another one, who has better insight into the modified code, has a closer look before we merge this PR.

maehne commented:
@davidhutchens and @kevinawalsh: Could you maybe have a look at this PR as well?

davidhutchens reviewed (changes requested):
This PR does clean up the code, using a java Queue rather than building one.

Note that the PR only changes the displayed actual clock frequency (when active) and does not change the way the simulator deals with clock ticks or the actual speed of the simulation. I do not suggest the scope be expanded, I am just clarifying the scope of the PR.


Bugs:
If you run at high frequency and then change frequency to something very low (like .5Hz) the first displayed value is NaN.

If you stop the clock (Auto-click disabled) and then restart it, the clock rate drops a lot and slowly moves back up. Likely, the old ticks are counted and the duration of time for the stopped ticks shows as a reduced rate. Restarting the clock should reset the TickCounter (and the SmoothingFilter).

Specifically, places the old code called updateSimulator, you now just assign the simulator and do not clear the SmoothingFilter. That may lead to miscalculated frequencies.


Code Style:
Your new files use cr-lf line endings. Our files use the Unix standard lf without the cr. Please remove the cr character from the end of each line.

Our code base uses camel case without underscores for variable names. You have underscores to start the instance variables of your classes. Please remove that underscore. Note that your constants with all upper-case and imbedded underscores is consistent with our style.

Most of your local variables are not changed and could be declared final.

Please fix the bugs and clean up the code style.

maehne commented:
Thanks for the thorough review @davidhutchens!

maehne reviewed (changes requested):
@alexandre-lecoq: Please address the issues raised by @davidhutchens!

alexandre-lecoq commented:
> If you run at high frequency and then change frequency to something very low (like .5Hz) the first displayed value is NaN.
Not reproduced, but I removed the smoothing filter, it could have been the cause for that.

> If you stop the clock (Auto-click disabled) and then restart it, the clock rate drops a lot and slowly moves back up. Likely, the old ticks are counted and the duration of time for the stopped ticks shows as a reduced rate. Restarting the clock should reset the TickCounter (and the SmoothingFilter).
Fixed

> Specifically, places the old code called updateSimulator, you now just assign the simulator and do not clear the SmoothingFilter. That may lead to miscalculated frequencies.
Fixed

> Your new files use cr-lf line endings. Our files use the Unix standard lf without the cr. Please remove the cr character from the end of each line.
Files removed. Basic smoothing was enough.

> Our code base uses camel case without underscores for variable names
Fixed

> Most of your local variables are not changed and could be declared final.
Fixed

Also, I moved to System.nanoTime() rather than System.currentTimeMillis(), since there's a tick frequency option at 2048 KHz which corresponds to a period of 488.3 nanoseconds (well under 1 millisecond), I suppose it won't hurt.

davidhutchens commented on src/main/java/com/cburch/logisim/gui/main/TickCounter.java:
```diff
@@ -13,172 +13,111 @@
+  private long elapsedTimeSinceLastUnitUpdate = 0;
+  static final int NANOSECONDS_UNIT_UPDATE_THRESHOLD = 500000000;
+  static final int TICKS_THRESHOLD_BEFORE_HISTORY_WEIGHT_REDUCTION = 1000000000;
+  static final int WEIGHT_REDUCTION_TICKS_COUNT = TICKS_THRESHOLD_BEFORE_HISTORY_WEIGHT_REDUCTION / 2;
```
I recommend using the number of seconds before weight reduction. You can calculate the ticksThreshold and weightReductionTicks from the number of seconds and the computed fullTicksPerSecond. A value of around 5 seconds worked in my playing with it.

The value you have is so large that you likely never drop earlier things. At 100 ticks per second, it takes a long time to get to a billion ticks. And you really want a larger number of ticks when the clock is running fast and a smaller number when it is ticking slow.

I have sample hardware simulations that can change the speed during a run because parts of the hardware shut down (i.e. don't change state). So I can test it on that. It should change the reported speed as the simulation changes how fast it is actually running.

alexandre-lecoq replied:
It was a mistake I introduced when switching to nanoseconds.
I was supposed to be 1000.
Hopefully, it'll give a better feeling like that.

davidhutchens reviewed (changes requested):
I really like this idea for implementation. It is much simpler than what we have, or your first attempt to improve it. I made some specific comments in the code that I think will make it better.

By the way, you do not need to force-push an update. You can just add a commit to your branch when you address these suggestions.

alexandre-lecoq commented on src/main/java/com/cburch/logisim/gui/main/TickCounter.java:
```diff
@@ -13,172 +13,111 @@
+    formatter = new DecimalFormat("0.00");
+    formatter.setRoundingMode(RoundingMode.HALF_UP);
   }
+  // XXX: use nanoTime()
```
Fixed

alexandre-lecoq commented on src/main/java/com/cburch/logisim/gui/main/TickCounter.java:
```diff
@@ -13,172 +13,111 @@
+    if (Double.isNaN(fullCyclesPerSeconds)) {
+      return "If you see this message, there was a problem computing the running frequency.";
+    }
+
```
Fixed

alexandre-lecoq commented:
> By the way, you do not need to force-push an update. You can just add a commit to your branch when you address these suggestions.

Yes, I forgot again, I got new habits after starting using gerrit.

R3dst0ne commented:
There are huge numbers in multiple places. Using Scientific notation might help with readability and decrease the chance of errors by typing the wrong amount of 0's.

MarcinOrlowski reviewed (changes requested):
As per @R3dst0ne comment, please do change the code and replace `500000000` (multiple places) by less error-prone notation scientific notation `5E+8` (you can put unfolded value in comment if you really wish).

alexandre-lecoq commented:
I removed all those big values completely and only kept a unique constant with thousands separators
`static final int NANOSECONDS_PER_SECONDS = 1_000_000_000;`
Does that address the issue ?
I'm not sure it really needs to be a double.

R3dst0ne commented:
The unique constants are a good solution.
Still, I would prefer to use scientific notation in the definitions.

MarcinOrlowski commented:
@R3dst0ne we already do use `000_000_000` notation so I'd leave it as it as this approach also helps.

davidhutchens reviewed (approved):
I think the integer constants with underscores is better than floating point constants in scientific notation for variables declared as integers. So I would leave this as is.

This implementation looks clean to me. I tried it with some LE files and it has a similar feel to the original.

Thank you @alexandre-lecoq for your contribution.

MarcinOrlowski commented on src/main/java/com/cburch/logisim/gui/main/TickCounter.java:
```diff
@@ -13,172 +13,107 @@
+  private double previousFrequency = 0.0;
+  private long elapsedTimeSinceLastUnitUpdate = 0;
+  static final int NANOSECONDS_PER_SECONDS = 1_000_000_000;
+  static final int NANOSECONDS_UNIT_UPDATE_THRESHOLD = NANOSECONDS_PER_SECONDS / 2;
```
I'm wondering if `UNIT_UPDATE_THRESHOLD_IN_NANOSECONDS` is not better name for this const. Not crucial, yet `ns` thing is least important here so it should not be first word in the constant name (IMHO). Leaving final decision to you though.

alexandre-lecoq replied:
Done

MarcinOrlowski commented on src/main/java/com/cburch/logisim/gui/main/TickCounter.java:
```diff
@@ -13,172 +13,107 @@
+    // It provides a nicer effect at low frequencies, and doesn't hurt at high frequencies.
+    if (simulator != null) {
+      final var tickPeriodNanoseconds = NANOSECONDS_PER_SECONDS / simulator.getTickFrequency();
+      tickCount = 12;
```
Please add comment why `12`

alexandre-lecoq replied:
Done

maehne reviewed (approved):
Looks good to me, but it should get reviewed and tested also by @BFH-ktt1 before merging this PR. Thanks for your effort in polishing this PR @alexandre-lecoq!

BFH-ktt1 commented:
I tested and it looks good to me, I would like to have this in V3.8.0. @MarcinOrlowski : could you deblock?
